New chromium rules

thypon opened the pull request.

goodov commented on assets/opengrep_rules/client/cryptography-random-usage.cpp:
```diff
@@ -0,0 +1,86 @@
+  void AcceptableUsage() {
+    // SHOULD NOT TRIGGER: Non-crypto random usage patterns
+    std::random_device rd;
+    std::mt19937 gen(rd());
```
this should be banned. random in Chromium should be taken either from `crypto/` or from `base/`.
> Do not use any random number engines or generators from `<random>`. Instead, use `base::RandomBitGenerator`. (You may use the distributions from `<random>`.)

https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/c++-features.md#engines-and-generators-from-random_banned

thypon replied:
Solved in 4b975d7e80af35a8217518afe07327edb43177cf

goodov commented on assets/semgrep_rules/client/sensitive-data-access-validation.yaml:
```diff
@@ -0,0 +1,51 @@
+              - pattern: GetStoredPasswords(...)
+              - pattern: SavePassword($CREDENTIAL, ...)
+              - pattern: GetIncognitoProfile()
+              - pattern: IsOffTheRecord()
```
`IsOffTheRecord()` check may be too noisy.

thypon replied:
Removed in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/webui-origin-validation.yaml:
```diff
@@ -0,0 +1,43 @@
+    patterns:
+      - patterns:
+          - pattern-either:
+              - pattern-regex: void\s+Handle\w*\([^)]*std::string[^)]*\)
```
this one had all kinds of false positives, a method named `Handle*` might be any number of different things that have no relation to this at all. I think maybe you're generalizing some specific case here that really doesn't fit.

thypon replied:
removed in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/weakptr-factory-placement.cpp:
```diff
@@ -0,0 +1,57 @@
+  int data_;
+  std::string name_;
+  bool enabled_;
+};
```
please check for missing eol

thypon replied:
Fixed in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/service-worker-origin-validation.yaml:
```diff
@@ -0,0 +1,38 @@
+    patterns:
+      - patterns:
+          - pattern-either:
+              - pattern: RegisterServiceWorker($URL, $SCOPE, ...)
```
not sure what this is intended to match? CPP code for creating service workers is `ServiceWorkerContext`

bridiver replied:
I think you're looking at methods like `ChromeAutocompleteProviderClient::StartServiceWorker`, but people aren't necessarily going to follow that convention and internally it get `content::ServiceWorkerContext` from the partition

thypon replied:
Added in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/service-worker-origin-validation.cpp:
```diff
@@ -0,0 +1,74 @@
+  void StartServiceWorker(const GURL& url, base::OnceCallback<void()> callback) {}
+  void RefreshServiceWorkerCache() {}
+  void UpdateServiceWorkerRegistration() {}
+};
```
eol

thypon replied:
Fixed in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/sensitive-data-access-validation.yaml:
```diff
@@ -0,0 +1,51 @@
+              - pattern: SetCookie($URL, $COOKIE, ...)
+              - pattern: GetStoredPasswords(...)
+              - pattern: SavePassword($CREDENTIAL, ...)
+              - pattern: GetIncognitoProfile()
```
I think you're looking for `GetOffTheRecordProfile`? There is no such method as `GetIncognitoProfile`

thypon replied:
Added in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/runloop-lifecycle-safety.yaml:
```diff
@@ -0,0 +1,30 @@
+    severity: WARNING
+    patterns:
+      - pattern-either:
+          - pattern: $MEMBER = $LOOP.QuitClosure()
```
this is already handled by presbumit

thypon replied:
Removed in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/process-privilege-validation.yaml:
```diff
@@ -0,0 +1,22 @@
+      compromised renderers to access resources they shouldn't be able to reach.
+    severity: ERROR
+    patterns:
+      - pattern-regex: void\s+\w*(Commit|Navigate|Access)\w*\([^)]*\)|bool\s+\w*Load\w*\([^)]*RenderProcessHost[^)]*\)
```
These are going to be way too noisy and catch all kinds of things that are unrelated to this

thypon replied:
Removed in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/content-script-injection-validation.yaml:
```diff
@@ -0,0 +1,50 @@
+      category: security
+    languages: [cpp, c]
+    message: |
+      Content script injection requires security review. Scripts injected into
```
content script injection is only done by extensions so isn't relevant here, I think we already have a rule that covers executing scripts?

thypon replied:
Removed in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/content-script-injection-validation.yaml:
```diff
@@ -0,0 +1,50 @@
+          - pattern-either:
+              - pattern: ExecuteScript($SCRIPT, ...)
+              - pattern: InjectScript($CODE, ...)
+              - pattern: AddContentScript($SCRIPT_NAME, ...)
```
this method only exists in tests

thypon replied:
Removed from this patchset 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/content-script-injection-validation.yaml:
```diff
@@ -0,0 +1,50 @@
+              - pattern: ExecuteScript($SCRIPT, ...)
+              - pattern: InjectScript($CODE, ...)
+              - pattern: AddContentScript($SCRIPT_NAME, ...)
+              - pattern: RegisterContentScript($DETAILS, ...)
```
this doesn't exist at all and per above isn't relevant because content scripts are handled entirely within extensions

thypon replied:
Removed from this patchset 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/semgrep_rules/client/crash-vs-dump-without-crashing.yaml:
```diff
@@ -0,0 +1,27 @@
+      - cpp
+      - c
+    message: |
+      Direct use of base::debug::DumpWithoutCrashing() should be replaced with
```
This is not accurate, it says "prefer", not "should" and there are cases where this won't necessarily be the case because we're using the dump to debug something that won't necessarily become a CHECK.

thypon replied:
Changed the wording in 4b975d7e80af35a8217518afe07327edb43177cf

bridiver commented on assets/opengrep_rules/client/cryptography-random-usage.yaml:
```diff
@@ -0,0 +1,36 @@
+    severity: WARNING
+    patterns:
+      - pattern-either:
+          - pattern: rand()
```
is semgrep not smart enough to recognize this with `using`?

bridiver commented on assets/opengrep_rules/client/mojo-security-validation.yaml:
```diff
@@ -0,0 +1,41 @@
+      category: security
+    languages: [cpp, c]
+    message: |
+      Mojo interface methods that receive privilege-presuming data (like file paths,
```
this references mojo interfaces, but this rule applies to cpp code in general and will be way too noisy

thypon replied:
I tried to narrow them down here https://github.com/brave/security-action/pull/855/commits/f03528b8734958fa38d243ea2785a93a1efb9a7b . Let me know if not enough, so I can remove these

bridiver replied:
if we're going to do this I think it makes more sense to target the mojom files and not the cpp files because these patterns are still way to general for the thing they are trying to target

bridiver commented on assets/semgrep_rules/client/network-request-validation.yaml:
```diff
@@ -0,0 +1,53 @@
+      category: security
+    languages: [cpp, c]
+    message: |
+      Network requests and URL modifications require security review. Ensure
```
this is going to miss a lot of things because of various wrappers around `SimpleURLLoader`. Looking for `net::DefineNetworkTrafficAnnotation` is probably better here

thypon replied:
I'm not sure this would cut it though, the fix it's too simplistic 4b975d7e80af35a8217518afe07327edb43177cf

thypon replied:
https://github.com/brave/security-action/pull/855/commits/f03528b8734958fa38d243ea2785a93a1efb9a7b I still feel it's too simplistic

bridiver commented on assets/semgrep_rules/client/weakptr-factory-placement.yaml:
```diff
@@ -0,0 +1,20 @@
+      category: security
+    languages: [cpp, c]
+    message: |
+      WeakPtrFactory should be the last member variable in the class to ensure
```
This is not needed, the build will fail

thypon replied:
https://github.com/brave/security-action/pull/855/commits/f03528b8734958fa38d243ea2785a93a1efb9a7b Removed

cdesouza-chromium commented on assets/opengrep_rules/client/container-performance-choice.yaml:
```diff
@@ -0,0 +1,26 @@
+      category: correctness
+    languages: [cpp, c]
+    message: |
+      Consider using base::flat_map/base::flat_set instead of std::map/std::set
+      for small containers or write-once scenarios. Chromium guidelines suggest:
+      - Use base::flat_map for small collections (< ~100 elements)
+      - Use base::flat_map for write-once, read-many scenarios
+      - Only use std::map for large containers with frequent writes
+      - Use std::map if you need pointer stability
```
I don't think this advice is correct. For reference reference, there's a detailed guideline over these types of containers in https://chromium.googlesource.com/chromium/src/+/refs/heads/main/base/containers/README.md.

These days, `std::map` and `std::unordered_map` are nearly always not desired. With `base::flat_map` we have O(n) complexity on lookups too, so that's also something that the more modern hash maps with prefetch strategies can have an edge depending on the scenario.

I also wonder if we this alert will be too chatty.

thypon replied:
> If you require pointer stability (on either the key or value) then std::map and std::set are the also the best choices.

these are summarization of what is contained in these documents btw, that precise line is contained in that README.md

cdesouza-chromium replied:
Let me clarify what I mean.


>Map and set selection
>Usage advice
>If you just need a generic map or set container without any additional properties then prefer to use absl::flat_hash_map and absl::flat_hash_set. These are versatile containers that have good performance on both large and small sized data.

So the first advice in that document is to use `absl` containers. The excerpt above makes no mentions of these containers, or of the preference for them. In code review I have already been shifting my advice away from the use of standards containers (e.g. `std::map`) based on this most recent guidance.

There's also a lot of upstream work doing the same. Heres an example: https://issues.chromium.org/issues/441179972

I just feel that this is very nuanced, and I'm also concerned how people may feel about these alerts just because they've used a container.

Will this particular type of code trigger this alert:

```
void Foo(std::set<std::string_view>& s);
```

If so, then this is quite prohibitive because people will be inundated with alerts for things they are just passing along.

thypon replied:
> Will this particular type of code trigger this alert:

it does not look to trigger

bridiver commented on assets/opengrep_rules/client/sensitive-data-access-validation.yaml:
```diff
@@ -0,0 +1,50 @@
+              - pattern: GetStoredPasswords(...)
+              - pattern: SavePassword($CREDENTIAL, ...)
+              - pattern: GetOffTheRecordProfile()
+              - pattern: GetPrivateBrowsingData(...)
```
this pattern does not match anything

bridiver commented on assets/opengrep_rules/client/sensitive-data-access-validation.yaml:
```diff
@@ -0,0 +1,50 @@
+              - pattern: SavePassword($CREDENTIAL, ...)
+              - pattern: GetOffTheRecordProfile()
+              - pattern: GetPrivateBrowsingData(...)
+              - pattern: AccessHistoryData($QUERY, ...)
```
same here, if you want to match history then it should be HistoryService

bridiver commented on assets/opengrep_rules/client/sensitive-data-access-validation.yaml:
```diff
@@ -0,0 +1,50 @@
+          - pattern-either:
+              - pattern: GetCookies($URL, ...)
+              - pattern: SetCookie($URL, $COOKIE, ...)
+              - pattern: GetStoredPasswords(...)
```
this is also not a valid pattern that will match access to password data. I don't know off hand what would correctly target the password database

bridiver commented on assets/opengrep_rules/client/sensitive-data-access-validation.yaml:
```diff
@@ -0,0 +1,50 @@
+    patterns:
+      - patterns:
+          - pattern-either:
+              - pattern: GetCookies($URL, ...)
```
this is not going to match actual access to cookies. Not sure exactly what you're trying to catch here? Generally in the browser process you would access cookies through network headers or RestrictedCookieManager

thypon committed:
Add check-vs-dcheck opengrep rule for Chromium

Flag DCHECK usage for security-critical invariants that should use CHECK
instead to prevent bypasses in release builds.

thypon committed:
Add container-performance-choice opengrep rule for Chromium

Suggest base::flat_map over std::map for better performance with small
datasets, following Chromium performance guidelines.

thypon committed:
Add crash-vs-dump-without-crashing opengrep rule for Chromium

Recommend DUMP_WILL_BE_CHECK macro instead of direct crashes for better
crash diagnostics and debugging in production.

thypon committed:
Add cryptography-random-usage opengrep rule for Chromium

Detect weak random number generation (rand/srand) and ensure crypto::RandBytes
is used for security-sensitive contexts.

thypon committed:
Add mojo-security-validation opengrep rule for Chromium

Enforce privilege validation in Mojo IPC handlers to prevent privilege
escalation through IPC message spoofing.

thypon committed:
Add network-request-validation opengrep rule for Chromium

Validate network requests containing user data to prevent SSRF and ensure
proper traffic annotation.

thypon committed:
Add sensitive-data-access-validation opengrep rule for Chromium

Check sensitive data access controls to ensure proper authorization before
accessing passwords, cookies, and other private user data.

thypon committed:
Add service-worker-origin-validation opengrep rule for Chromium

Validate service worker origins to prevent unauthorized service worker
registration and scope manipulation attacks.
